### PR TITLE
【KernelGen】Add scatter_reduce_ operator

### PR DESCRIPTION
**Pull request by Schopenhauer-loves-Hegel**
### PR Category
Operator

### Type of Change
New Feature

### Description
Add `scatter_reduce_` operator implementation with Triton kernel.

- Implementation mode: `manual_kernel`
- Accuracy test: 120/120 passed

### Issue
N/A

### Progress

- [ ] Change is properly reviewed (1 reviewer required, 2 recommended).
- [ ] Change is responded to an issue.
- [x] Change is fully covered by a UT.

### Performance
**torch.float16**

| Shape | Torch Latency (ms) | Gems Latency (ms) | Speedup |
|-------|-------------------:|-------------------:|--------:|
| [1024, 1024] | 0.0413 | 0.0633 | 0.653 |
| [4096, 4096] | 0.4259 | 1.1239 | 0.379 |
| [1024, 1024] | 0.0686 | 0.0896 | 0.765 |
| [1024, 1024] | 0.0565 | 0.0556 | 1.017 |
| [4096, 4096] | 0.9758 | 0.8354 | 1.168 |
| [1024, 1024] | 0.0567 | 0.0550 | 1.030 |
| [4096, 4096] | 0.9740 | 0.8335 | 1.169 |
| [1024, 1024] | 0.0932 | 0.2843 | 0.328 |

**torch.float32**

| Shape | Torch Latency (ms) | Gems Latency (ms) | Speedup |
|-------|-------------------:|-------------------:|--------:|
| [1024, 1024] | 0.0377 | 0.0364 | 1.037 |
| [4096, 4096] | 0.3931 | 0.3965 | 0.991 |
| [1024, 1024] | 0.0671 | 0.0874 | 0.768 |
| [1024, 1024] | 0.0595 | 0.0557 | 1.068 |
| [4096, 4096] | 0.8990 | 0.8166 | 1.101 |
| [1024, 1024] | 0.0602 | 0.0564 | 1.066 |
| [4096, 4096] | 0.8990 | 0.8167 | 1.101 |
| [4096, 4096] | 1.1928 | 1.0919 | 1.092 |

**Overall: median speedup = 1.034x, mean speedup = 0.921x** (16 data points)

---
_Generated by auto_gen tool with Claude Code_

### CODE REVIEW

**Comment by tengqm** [comment]
@Schopenhauer-loves-Hegel There are some linter nits to be fixed.

**Comment by factnn** [comment]
## Benchmark Results for `scatter_reduce_` operator

### Summary
- **sum**: Good speedup on most sizes (0.98x-1.02x)
- **prod**: Mixed results, slower on large tensors (0.48x-2.17x)
- **amax/amin**: Near parity with PyTorch (0.85x-1.55x)
- **mean**: Competitive performance (0.86x-1.16x)

### scatter_reduce_.sum (float16)
| Torch (ms) | Gems (ms) | Speedup | Torch GBPS | Gems GBPS | Size |
|-----------|----------|---------|-----------|----------|------|
| 0.020 | 0.014 | 1.429 | 2.460 | 3.515 | [64,64] |
| 0.020 | 0.015 | 1.333 | 39.575 | 52.767 | [256,256] |
| 0.055 | 0.056 | 0.982 | 227.419 | 223.327 | [1024,1024] |
| 0.398 | 0.398 | 1.000 | 675.411 | 675.194 | [4096,4096] |
| 1.589 | 1.559 | **1.019** | 675.901 | 688.826 | [1024,65536] |
| 0.074 | 0.067 | **1.100** | 556.643 | 612.440 | [10000,256] |
| 15.43 | 15.09 | **1.023** | 679.740 | 695.069 | [10000,65536] |

### scatter_reduce_.prod (float16)
| Torch (ms) | Gems (ms) | Speedup | Torch GBPS | Gems GBPS | Size |
|-----------|----------|---------|-----------|----------|------|
| 0.048 | 0.022 | **2.166** | 1.017 | 2.204 | [64,64] |
| 0.050 | 0.025 | **1.973** | 15.825 | 31.227 | [256,256] |
| 0.074 | 0.091 | 0.819 | 169.490 | 138.897 | [1024,1024] |
| 0.972 | 1.526 | 0.637 | 207.205 | 131.907 | [4096,4096] |
| 4.565 | 7.896 | 0.578 | 176.403 | 101.994 | [1024,65536] |
| 0.154 | 0.202 | 0.759 | 200.125 | 151.935 | [10000,256] |
| 45.38 | 80.94 | 0.561 | 173.292 | 97.163 | [10000,65536] |

### scatter_reduce_.amax (float16)
| Torch (ms) | Gems (ms) | Speedup | Torch GBPS | Gems GBPS | Size |
|-----------|----------|---------|-----------|----------|------|
| 0.021 | 0.014 | **1.552** | 2.324 | 3.606 | [64,64] |
| 0.020 | 0.015 | **1.283** | 39.575 | 50.777 | [256,256] |
| 0.056 | 0.056 | 0.997 | 225.856 | 225.210 | [1024,1024] |
| 0.828 | 0.838 | 0.988 | 243.162 | 240.224 | [4096,4096] |
| 4.046 | 4.753 | 0.851 | 210.838 | 179.472 | [1024,65536] |
| 0.127 | 0.133 | 0.955 | 322.835 | 308.242 | [10000,256] |
| 42.51 | 49.90 | 0.852 | 185.006 | 157.577 | [10000,65536] |

### scatter_reduce_.amin (float16)
| Torch (ms) | Gems (ms) | Speedup | Torch GBPS | Gems GBPS | Size |
|-----------|----------|---------|-----------|----------|------|
| 0.021 | 0.014 | **1.500** | 2.324 | 3.486 | [64,64] |
| 0.020 | 0.015 | **1.283** | 39.575 | 50.777 | [256,256] |
| 0.056 | 0.056 | 1.000 | 225.856 | 225.856 | [1024,1024] |
| 0.828 | 0.838 | 0.988 | 243.162 | 240.224 | [4096,4096] |
| 4.046 | 4.753 | 0.851 | 210.838 | 179.472 | [1024,65536] |
| 0.127 | 0.133 | 0.955 | 322.835 | 308.242 | [10000,256] |
| 42.51 | 49.90 | 0.852 | 185.006 | 157.577 | [10000,65536] |

### scatter_reduce_.mean (float16)
| Torch (ms) | Gems (ms) | Speedup | Torch GBPS | Gems GBPS | Size |
|-----------|----------|---------|-----------|----------|------|
| 0.021 | 0.018 | **1.167** | 2.324 | 2.712 | [64,64] |
| 0.020 | 0.017 | **1.176** | 39.575 | 46.559 | [256,256] |
| 0.056 | 0.064 | 0.875 | 225.856 | 197.624 | [1024,1024] |
| 0.828 | 0.959 | 0.863 | 243.162 | 209.929 | [4096,4096] |
| 4.046 | 4.688 | 0.863 | 210.838 | 181.993 | [1024,65536] |
| 0.127 | 0.148 | 0.858 | 322.835 | 276.993 | [10000,256] |
| 42.51 | 49.37 | 0.861 | 185.006 | 159.283 | [10000,65536] |

**Overall**: sum performs best with near-parity or slight speedup. prod shows slowdown on large tensors. amax/amin/mean are competitive.

**Comment by tengqm** [comment]
It is weird that the pytest options are registered twice ...

File: src/flag_gems/__init__.py [hunk 1]
```diff
@@ -391,6 +391,7 @@ def torch_ge(v):
     ("scatter_.reduce", scatter_),
     ("scatter_.src", scatter_),
     ("scatter_add_", scatter_add_),
+    ("scatter_reduce_.two", scatter_reduce_),
```
Comment (tengqm) [inline src/flag_gems/__init__.py:420]: This is a very special condition ... we cannot use `scatter_reduce` to identify this operator, because it conflicts with line 389; we cannot use `scatter_reduce_` either, because that will conflict with line 391;
the only choice, is to use `scatter_reduce_two_` as the identifier for this operator.

That means, we need to mark all accuracy tests and benchmarks with `@pytest.mark.scatter_reduce_two_`.

File: tests/test_scatter_reduce_.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+import pytest
+import torch
+
+import flag_gems
+
+from .accuracy_utils import gems_assert_close, to_reference
+from .conftest import QUICK_MODE
+
+
+@pytest.mark.scatter_reduce_
```
Comment (tengqm) [inline tests/test_scatter_reduce_.py]:
```suggestion
@pytest.mark.scatter_reduce_two_
```

File: tests/test_scatter_reduce_two.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
+import random
```
Comment (tengqm) [inline tests/test_scatter_reduce_two.py:1]: Rename the file to `test_scatter_reduce_two.py`.

File: tests/test_scatter_reduce_.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+@pytest.mark.parametrize(
+    "src_shape", [(32, 8, 4)] if QUICK_MODE else [(128, 16, 4), (256, 32, 8)]
+)
+@pytest.mark.parametrize(
+    "inp_shape", [(64, 16, 8)] if QUICK_MODE else [(512, 128, 32), (1024, 64, 16)]
+)
+@pytest.mark.parametrize("dim", [0, 1, 2])
+@pytest.mark.parametrize("dtype", [torch.float16, torch.float32])
+@pytest.mark.parametrize("reduce", ["sum", "prod", "amax", "amin", "mean"])
+def test_scatter_reduce_(src_shape, inp_shape, dim, dtype, reduce):
```
Comment (tengqm) [inline tests/test_scatter_reduce_.py]:
```suggestion
def test_scatter_reduce_two_(src_shape, inp_shape, dim, dtype, reduce):
```

File: benchmark/test_scatter_reduce_perf.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        op_name="scatter_reduce_.amin",
+        torch_op=torch.Tensor.scatter_reduce_,
+        input_fn=scatter_reduce_input_fn_factory("amin"),
+        dtypes=[torch.float16, torch.float32],
+        is_inplace=True,
+    )
+    bench.run()
+
+
+@pytest.mark.scatter_reduce_
```
Comment (tengqm) [inline benchmark/test_scatter_reduce_perf.py]:
```suggestion
@pytest.mark.scatter_reduce_two_
```

File: benchmark/test_scatter_reduce_perf.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+def test_perf_scatter_reduce_mean():
```
Comment (tengqm) [inline benchmark/test_scatter_reduce_perf.py]:
```suggestion
def test_scatter_reduce_two_inplace_mean():
```

File: benchmark/test_scatter_reduce_perf.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+        op_name="scatter_reduce_.amax",
+        torch_op=torch.Tensor.scatter_reduce_,
+        input_fn=scatter_reduce_input_fn_factory("amax"),
+        dtypes=[torch.float16, torch.float32],
+        is_inplace=True,
+    )
+    bench.run()
+
+
+@pytest.mark.scatter_reduce_
```
Comment (tengqm) [inline benchmark/test_scatter_reduce_perf.py]:
```suggestion
@pytest.mark.scatter_reduce_two_
```

File: benchmark/test_scatter_reduce_perf.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+def test_perf_scatter_reduce_amin():
```
Comment (tengqm) [inline benchmark/test_scatter_reduce_perf.py]:
```suggestion
def test_scatter_reduce_two_inplace_amin():
```

File: benchmark/test_scatter_reduce_perf.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+        op_name="scatter_reduce_.sum",
+        torch_op=torch.Tensor.scatter_reduce_,
+        input_fn=scatter_reduce_input_fn_factory("sum"),
+        dtypes=[torch.float16, torch.float32],
+        is_inplace=True,
+    )
+    bench.run()
+
+
+@pytest.mark.scatter_reduce_
```
Comment (tengqm) [inline benchmark/test_scatter_reduce_perf.py]:
```suggestion
@pytest.mark.scatter_reduce_two_
```

File: benchmark/test_scatter_reduce_perf.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+def test_perf_scatter_reduce_amax():
```
Comment (tengqm) [inline benchmark/test_scatter_reduce_perf.py]:
```suggestion
def test_scatter_reduce_two_inplace_amax():
```

File: benchmark/test_scatter_reduce_perf.py [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+
+class ScatterReduceBenchmark(GenericBenchmark2DOnly):
+    def set_more_metrics(self):
+        return ["gbps"]
+
+    def get_gbps(self, args, latency):
+        return scatter_reduce_gbps(args, latency)
+
+
+@pytest.mark.scatter_reduce_
```
Comment (tengqm) [inline benchmark/test_scatter_reduce_perf.py]:
```suggestion
@pytest.mark.scatter_reduce_two_
```

File: benchmark/test_scatter_reduce_perf.py [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+def test_perf_scatter_reduce_sum():
```
Comment (tengqm) [inline benchmark/test_scatter_reduce_perf.py]:
```suggestion
def test_scatter_reduce_two_inplace_sum():
```

**Review by tengqm — APPROVED** [review]

**Comment by factnn** [comment]
## UT Results

Branch rebased to upstream/master and tests passed:

```
python -m pytest tests/ -m scatter_reduce_ --ref cpu -x -q
120 passed
```

**Review by bin913 — APPROVED** [review]
